cmake/GzUtils.cmake

```diff
@@ -1944,6 +1944,161 @@ macro(gz_build_tests)
 
 endmacro()
 
+#################################################
+# gz_build_examples(
+#     SOURCE_DIR <source_dir>
+#     BINARY_DIR <binary_dir>
+#     [DEPENDENCIES <project_dependencies>]
+#
+# Requires a CMakeLists.txt file to be in SOURCE_DIR that acts
+# Build examples for a Gazebo project.
+# as a top level project.
+#
+# This generates two test targets
+# * EXAMPLES_Configure_TEST - Equivalent of calling "cmake .." on
+#                             the examples directory
+#
+# * EXAMPLES_Build_TEST - Equivalent of calling "make" on the
+#                         examples directory
+#
+# These tests are run during "make test" or can be run specifically
+# via "ctest -R EXAMPLES_ -V"
+#
+# Arguments are as follows:
+#
+# SOURCE_DIR: Required. Path to the examples folder.
+#             For example ${CMAKE_CURRENT_SOURCE_DIR}/examples
+#
+# BINARY_DIR: Required. Path to the output binary folder
+#             For example ${CMAKE_CURRENT_BINARY_DIR}/examples
+#
+# DEPENDENCIES:
+#     Optional. Additional Gazebo library dependencies for this example
+#     This is in addition to things that the project library already depends on.
+#     For example: gz-utils2 gz-common5
+
+macro(gz_build_examples)
+  #------------------------------------
+  # Define the expected arguments
+  set(options)
+  set(oneValueArgs SOURCE_DIR BINARY_DIR)
+  set(multiValueArgs DEPENDENCIES)
+
+  #------------------------------------
+  # Parse the arguments
+  _gz_cmake_parse_arguments(gz_build_examples "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+  set(BUILD_EXAMPLES_CMAKE "")
+  set(BUILD_EXAMPLES_INCLUDES "")
+
+  get_target_property(PROJECT_LIBRARY_DEPS ${PROJECT_LIBRARY_TARGET_NAME} INTERFACE_LINK_LIBRARIES)
+
+  # Below here we are generating a series of arguments to the cmake command that is to follow.
+  # Since we may be building in a colcon workspace or otherwise isolated, we are passing
```

Comment:
It's not clear to me why all of this is needed. I deleted `BUILD_EXAMPLES_CMAKE` and all the related code and the test in gz-math still works. I even added a dependency on `gz-utils2` in the gz-math example just to make sure. It kinda makes sense that it worked because the examples are all self contained and whatever you use to run them, be it colcon or isolated cmake, it would have the same requirements as the package that contains them. If you're able to build the package, I would think you would be able to build the examples without the need to pass resolved dependency directories.

Reply:
Now it's been a bit, but I seem to recall the issue as follows.  It's not when the examples depend on a package that the core is already depending on (eg math already exists on gz-utils).  The issue arises when examples depend on something that isn't part of the core package, such as an example depending on `gz-utils2::cli`.

Reply:
Ah, I remember now, this was extra work so that `make test` and `ctest` work directly from the `build` directory. An example that definitely fails without the dependencies (but passes with `colcon test`)

In gz-common without the dependencies set (these lines removed), it fails:

https://github.com/gazebosim/gz-common/blob/569513a14f703a16a5c7ddd2453c6ce037a03cb9/CMakeLists.txt#L169-L172

So I suppose is if the extra cmake boilerplate is worth it in order to support `make && make test` or `ninja && ninja test` (which is also how we invoke in github actions CI at the moment, iirc)

Reply:
Ah, I see the problem now as I'm testing the gz-common PR. The way `colcon test` works is by setting the `CMAKE_PREFIX_PATH` environment variable, so what if we did the same based on the environment variable set during `colcon build`? I tried doing that and it seems to work for both `colcon test` and working directly from the `build` directory.

```cmake

  set(gz_build_examples_CMAKE_PREFIX_PATH $ENV{CMAKE_PREFIX_PATH})
  string(REPLACE ":" ";" gz_build_examples_CMAKE_PREFIX_PATH ${gz_build_examples_CMAKE_PREFIX_PATH})
  list(APPEND gz_build_examples_CMAKE_PREFIX_PATH ${CMAKE_INSTALL_PREFIX})

 # then pass it to `add_test`
  add_test(
    NAME EXAMPLES_Configure_TEST
    COMMAND ${CMAKE_COMMAND} -G${CMAKE_GENERATOR}
                             --no-warn-unused-cli
                             -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
                             -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
                             -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
                             -DCMAKE_C_FLAGS=${CMAKE_C_FLAGS}
                             -DCMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS}
                             -S ${gz_build_examples_SOURCE_DIR}
                             -B ${gz_build_examples_BINARY_DIR}
                             "-DCMAKE_PREFIX_PATH=${gz_build_examples_CMAKE_PREFIX_PATH}"

Reply:
Oh, yeah, that should work and be much less code to maintain.  Don't know why I didn't think of that.